core/templates/dev/head/pages/exploration_editor/editor_tab/StateSolution.js

```diff
@@ -0,0 +1,214 @@
+ */
+
+oppia.controller('StateSolution', [
+  '$scope', '$rootScope', '$modal', 'editorContextService', 'alertsService',
+  'INTERACTION_SPECS', 'stateSolutionService', 'explorationStatesService',
+  'SolutionVerificationService', 'oppiaExplorationHtmlFormatterService',
+  'stateInteractionIdService', 'stateHintsService', 'UrlInterpolationService',
+  'SolutionObjectFactory', 'responsesService', 'explorationContextService',
+  'explorationWarningsService', 'INFO_MESSAGE_SOLUTION_IS_INVALID',
+  function(
+    $scope, $rootScope, $modal, editorContextService, alertsService,
+    INTERACTION_SPECS, stateSolutionService, explorationStatesService,
+    SolutionVerificationService, oppiaExplorationHtmlFormatterService,
+    stateInteractionIdService, stateHintsService, UrlInterpolationService,
+    SolutionObjectFactory, responsesService, explorationContextService,
+    explorationWarningsService, INFO_MESSAGE_SOLUTION_IS_INVALID) {
+    $scope.stateSolutionService = stateSolutionService;
+    $scope.inlineSolutionEditorIsActive = false;
+    $scope.correctAnswerEditorHtml = '';
+    $scope.SOLUTION_EDITOR_FOCUS_LABEL = (
+      'currentCorrectAnswerEditorHtmlForSolutionEditor');
+    explorationWarningsService.updateWarnings();
+    $scope.stateHintsService = stateHintsService;
+    $scope.correctAnswer = null;
+
+    $scope.isSolutionValid = function() {
+      return explorationStatesService.isSolutionValid(
+        editorContextService.getActiveStateName());
+    };
+
+    $scope.inlineSolutionEditorIsActive = false;
+    $scope.stateInteractionIdService = stateInteractionIdService;
+
+    $scope.correctAnswerEditorHtml = (
+        oppiaExplorationHtmlFormatterService.getInteractionHtml(
+          stateInteractionIdService.savedMemento,
+          explorationStatesService.getInteractionCustomizationArgsMemento(
+            editorContextService.getActiveStateName()),
+          $scope.SOLUTION_EDITOR_FOCUS_LABEL));
+
+    $scope.toggleInlineSolutionEditorIsActive = function() {
+      $scope.inlineSolutionEditorIsActive = (
+        !$scope.inlineSolutionEditorIsActive);
+    };
+
+    $scope.getSolutionSummary = function() {
+      var solution = stateSolutionService.savedMemento;
+      return solution.getSummary(
+        stateInteractionIdService.savedMemento,
+        responsesService.getAnswerChoices());
+    };
+
+    // This returns false if the current interaction ID is null.
+    $scope.isCurrentInteractionLinear = function() {
+      return (
+        stateInteractionIdService.savedMemento &&
+        INTERACTION_SPECS[stateInteractionIdService.savedMemento].is_linear);
+    };
+
+    $scope.openAddSolutionModal = function() {
+      alertsService.clearWarnings();
+      $rootScope.$broadcast('externalSave');
+      $scope.inlineSolutionEditorIsActive = false;
+
+      $modal.open({
+        templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+          '/pages/exploration_editor/editor_tab/add_solution_modal.html'),
```

Comment:
Ditto: add_or_update_solution_modal.html? Otherwise the reader is wondering why a correctAnswer is already supplied.